### PR TITLE
feat: Added the ability to filter the corresponding files based on the access_time, modify time, and change time of the file for statistics.

### PR DESCRIPTION
**Pull request by wugeer**
# Background

Filter by file age https://github.com/bootandy/dust/issues/340
Add a new feature to dust to filter files based on their last update time, last access time, and last modification time.          

In order to achieve this requirement, I referred to the input of the linux find command, and ultimately planned to implement the dust filter by file age function based on the input style of find.

image

I have a little question here, why does dust output files or directories with a file size of 0 by default? The dust tool is used to quickly find those files and directories in the specified directory that take up too much space, and those empty files or directories will only affect the look and feel. Originally, I have implemented filtering out those files or directories with a size of 0, but when I run cargo test When running the test, I found that some samples reported errors, so I canceled this part of the code, but I still wanted to retain the function of filtering out those files or directories with a size of 0. I hope this function can be further explored. thanks

# Full Changelogs

- feat: Added the ability to filter the corresponding files based on the access_time, modify time, and change time of the file for statistics

# Issue Reference

fix:  Filter by file age https://github.com/bootandy/dust/issues/340

# Test Result
dust help command output
image

Test Data

image

cargo test run test results

image

Adding --mtime or --atime or --ctime results in actual execution of dust and linux find with similar parameters.

image

image

image

image

image

### CODE REVIEW

**Comment by bootandy** [comment]
Apologies I've had a laptop failure. Don't want to risk looking at this on my work laptop for legal reasons. Will review this when I get a new machine.

File: completions/_dust
Comment (codeswhite) [inline completions/_dust:40]: Consider making the description more easily readable


```suggestion
'-M+[Sort by time modifed, include only files modified +/-<mtime> days before/after, days are rounded down, prefix \'\-\' for newer, \'\+\' for older. : : ' \
```

Comment (wugeer) [reply completions/_dust:40]: In my opinion, Filter by time mofied is better than Sort by time modifed because this parameter is used for filtering, not sorting. right? :)

Comment (codeswhite) [reply completions/_dust:40]: Sure, you can rephrase it as you wish, just pointed that its a bit too long :)

File: src/dir_walker.rs
Comment (codeswhite) [inline src/dir_walker.rs:194]: Consider changing `build_node` to accept a `walk_data` argument instead of individually 7 of `walk_data`'s fields

File: completions/_dust
Comment (wugeer) [reply completions/_dust]: Indeed, that description is a bit too long. I rephrased it . I will be very happy if you can help me review the code again. :)

File: src/dir_walker.rs
Comment (wugeer) [reply src/dir_walker.rs]: Indeed, you are right, I reorganized the code according to your ideas. I will be very happy if you can help me review the code again. :)

**Review by codeswhite — APPROVED** [review]
Looks good!
Should wait for @bootandy to review as well

I still suggest that after merging this we continue to work on a more robust implementation that will support all timeframes (like i explained in #340 )

**Comment by wugeer** [comment]
> Looks good! Should wait for @bootandy to review as well
>
> I still suggest that after merging this we continue to work on a more robust implementation that will support all timeframes (like i explained in #340 ) All timeframes (like I explained in #340).

That sounds like a good idea. I like it! I'll try it. ：)

**Comment by wugeer** [comment]
> Looks good! Should wait for @bootandy to review as well
>
> I still suggest that after merging this we continue to work on a more robust implementation that will support all timeframes (like i explained in #340 )

I have thought about it carefully these days. If I modify it according to your idea, I need to add a parameter to limit the upper and lower limits of the filtering time. Otherwise, there will be no equal scenarios. Although this method provides great flexibility, As for whether the tool itself needs such great flexibility, and whether adding this function will bring a lot of mental burden to the user, after all, everyone may have used it and are accustomed to the find command's -mtime, which is in units of days or - mmin is just filtering with two granularities, what do you think about it?> Looks good! Should wait for @bootandy to review as well
> 
> I still suggest that after merging this we continue to work on a more robust implementation that will support all timeframes (like i explained in #340 )

**Comment by wugeer** [comment]
> Looks good! Should wait for @bootandy to review as well
>
> I still suggest that after merging this we continue to work on a more robust implementation that will support all timeframes (like i explained in #340 )

I've been thinking seriously about this over the past few days. If we modify the tool according to your approach, we'll need to add a parameter to set the upper and lower limits of the filter time. Otherwise, the scenario for equal values would no longer exist. While this method provides great flexibility, I'm concerned about whether such a high level of flexibility is necessary for this tool. Adding this feature might place a significant cognitive burden on the users. After all, people are probably accustomed to using the find command with the -mtime option, which filters by days, or the -mmin option, which filters by minutes. What are your thoughts on this? @bootandy

File: completions/_dust.ps1
Comment (bootandy) [inline completions/_dust.ps1:51]: I'm not sure if we need all these - I like the idea of a including a filter that works with modified time. But 'access time' - would someone actually use that? I have to launch 'stat' to even see it.

File: src/dir_walker.rs
Comment (bootandy) [inline src/dir_walker.rs:187]: sensible change, there were too many params.

**Review by bootandy — CHANGES_REQUESTED** [review]
I like that you based the parameters for this work from find.

We seem to have a few cargo style problems. But overall I think this approach is correct.

File: completions/_dust.ps1
Comment (bootandy) [reply completions/_dust.ps1:51]: Ah we can leave it in if you'd like.

**Comment by wugeer** [comment]
> > I have a little question here, why does dust output files or directories with a file size of 0 by default? I have a little question, why does dust output files or directories with a file size of 0 by default?
>
> Just incase you are running on a very small folder. Some people seem to use dust as a sort of file-finder so I want to include them. Some people seem to use dust as a file finding tool, so I wanted to include them.

Yes, you are right! :)

**Comment by wugeer** [comment]
> I like that you based the parameters for this work from find. I like that you based the parameters for this work from find.
>
> We seem to have a few cargo style problems. But overall I think this approach is correct. We seem to have a few cargo style problems. But overall, I think this approach is the right one.

I'm fairly new to Rust , so my coding style may not be very good. If there is anything that needs to be modified, I hope you will give me some advice. thx!

File: completions/_dust.ps1
Comment (wugeer) [reply completions/_dust.ps1:51]: In my opinion, access time is an effective filtering tool when users want to determine whether a file has been accessed recently based on the access time to determine whether other people are using the file.

**Comment by bootandy** [comment]
>I'm fairly new to Rust , so my coding style may not be very good. If there is anything that needs to be modified, I hope you will give me some advice. thx!

Your coding style looks fine to me. The build however gave style errors ^ see the red 'X's in the build. 

On your cmd line:

Run 'cargo fmt' - this neatens things automatically

Run 'cargo clippy' - this is a linter which will suggest fixes and warns of likely bugs, so you'll have to review each one. 

you can see the list of errors from the online build here: https://github.com/bootandy/dust/actions/runs/9461768640/job/26334984807?pr=399

**Comment by wugeer** [comment]
> > I'm fairly new to Rust , so my coding style may not be very good. If there is anything that needs to be modified, I hope you will give me some advice. thx!
> 
> Your coding style looks fine to me. The build however gave style errors ^ see the red 'X's in the build.
> 
> On your cmd line:
> 
> Run 'cargo fmt' - this neatens things automatically
> 
> Run 'cargo clippy' - this is a linter which will suggest fixes and warns of likely bugs, so you'll have to review each one.
> 
> you can see the list of errors from the online build here: https://github.com/bootandy/dust/actions/runs/9461768640/job/26334984807?pr=399

Thank you for your guidance in helping me improve my code to meet the standards. I have resubmitted it, and all checks have passed. :)

**Comment by bootandy** [comment]
> this method provides great flexibility, I'm concerned about whether such a high level of flexibility is necessary for this tool. Adding this feature might place a significant cognitive burden on the users. After all, people are probably accustomed to using the find command with the -mtime option, which filters by days, or the -mmin option, which filters by minutes. What are your thoughts on this? @bootandy

I don't have strong opinions on this. But lets stabilise -mtime an -mmin for now. If we want to do the more flexible one we can do that in a followup. It definitely shouldn't go in this PR.

**Comment by bootandy** [comment]
Would you mind doing a rebase or merge - I've accepted a different PR and there are some conflicts ^ - thanks .

**Comment by wugeer** [comment]
> > this method provides great flexibility, I'm concerned about whether such a high level of flexibility is necessary for this tool. Adding this feature might place a significant cognitive burden on the users. After all, people are probably accustomed to using the find command with the -mtime option, which filters by days, or the -mmin option, which filters by minutes. What are your thoughts on this? @bootandy
> 
> I don't have strong opinions on this. But lets stabilise -mtime an -mmin for now. If we want to do the more flexible one we can do that in a followup. It definitely shouldn't go in this PR.

Indeed, I agree with you. :)

**Comment by wugeer** [comment]
> Would you mind doing a rebase or merge - I've accepted a different PR and there are some conflicts ^ - thanks .

Ok. I'm happy to do it! :)

File: src/config.rs
Comment (bootandy) [inline src/config.rs:13]: typo!

Comment (wugeer) [reply src/config.rs:13]: Sorry, due to my carelessness, I did not correct this spelling error in time. I will resubmit a PR for processing.

Comment (bootandy) [reply src/config.rs:13]: Not a problem! we all do it.

**Pull request merged**
